[lldb-dap] Refactoring IOStream into Transport handler.

ashgti opened the pull request:
Instead of having two discrete InputStream and OutputStream helpers, this merges the two into a unifed 'Transport' handler.

This handler is responsible for reading the DAP message headers, parsing the resulting JSON and converting the messages into `lldb_dap::protocol::Message`s for both input and output.

ashgti committed:
[lldb-dap] Refactoring IOStream into Transport handler.

Instead of having two discrete InputStream and OutputStream helpers,
this merges the two into a unifed 'Transport' handler.

This handler is responsible for reading the DAP message headers, parsing
the resulting JSON and converting the messages into
`lldb_dap::protocol::Message`s for both input and output.

ashgti commented:
#130090 is a follow up to add improved type handling to `RequestHandler` for encoding / decoding.

vogelsgesang commented on lldb/tools/lldb-dap/Transport.cpp:
```diff
@@ -0,0 +1,152 @@
+using namespace lldb_dap;
+using namespace lldb_dap::protocol;
+
+static Status ReadFull(IOObjectSP &descriptor, size_t length,
+                       std::string &text) {
```
can we make this an `Expected<std::string>` (or something similar) to avoid the output parameter here?

vogelsgesang reviewed (commented):
Why is this still marked as draft? It looks almost ready to be merged to me 🙂

vogelsgesang commented on lldb/tools/lldb-dap/DAP.cpp:
```diff
@@ -65,7 +66,7 @@ DAP::DAP(std::string name, llvm::StringRef path, std::ofstream *log,
          std::vector<std::string> pre_init_commands)
     : name(std::move(name)), debug_adapter_path(path), log(log),
-      input(std::move(input)), output(std::move(output)),
+      transport(this->name, std::move(input), std::move(output)),
```
nevermind, I initially misread the code

ashgti commented on lldb/tools/lldb-dap/Transport.h:
```diff
@@ -0,0 +1,56 @@
+  lldb_private::Status Read(std::ofstream *log, protocol::Message &M);
+
+  /// Closes the transport.
+  void CloseOutput();
```
Removed

ashgti commented on lldb/tools/lldb-dap/Transport.h:
```diff
@@ -0,0 +1,56 @@
+  void CloseOutput();
+
+private:
+  llvm::StringRef m_client_name;
```
Updated.

ashgti commented on lldb/tools/lldb-dap/DAP.cpp:
```diff
@@ -722,40 +683,9 @@ void DAP::SetTarget(const lldb::SBTarget target) {
-
-bool DAP::HandleObject(const llvm::json::Object &object) {
+bool DAP::HandleObject(const protocol::Message &M) {
+  llvm::json::Value v = toJSON(M);
```
Done.

vogelsgesang commented on lldb/tools/lldb-dap/JSONUtils.cpp:
```diff
@@ -1526,7 +1526,8 @@ static void addStatistic(lldb::SBTarget &target, llvm::json::Object &event) {
     FilterAndGetValueForKey(statistics, key, stats_body);
   }
-  event.try_emplace("statistics", std::move(stats_body));
+  llvm::json::Object body{{"$__lldb_statistics", std::move(stats_body)}};
+  event.try_emplace("body", std::move(body));
```
you could move this change into a separate commit which you land *before* this PR, not after

vogelsgesang reviewed (approved):
LGTM

Side note: Please don't amend your commits, but rather push multiple separate commits into your PR. That way, I can see the diff compared to the version which I previously reviewed

JDevlieghere commented on lldb/tools/lldb-dap/Transport.cpp:
```diff
@@ -0,0 +1,146 @@
+  return std::move(M);
+}
+
+lldb_private::Status Transport::Write(std::ofstream *log,
```
This should return an `llvm::Error`. 

```suggestion
llvm::Error Transport::Write(std::ofstream *log,
```

JDevlieghere commented on lldb/tools/lldb-dap/DAP.cpp:
```diff
@@ -838,19 +775,16 @@ llvm::Error DAP::Loop() {
-    if (status != lldb_dap::PacketStatus::Success) {
-      return llvm::createStringError(llvm::inconvertibleErrorCode(),
-                                     "failed to send packet");
+    auto next = transport.Read(log);
```
I think this goes beyond [LLVM's guidance of auto](https://llvm.org/docs/CodingStandards.html#use-auto-type-deduction-to-make-code-more-readable). It's not obvious from the right hand side what the type of `next` is.

JDevlieghere commented on lldb/tools/lldb-dap/DAP.cpp:
```diff
@@ -219,65 +220,27 @@ void DAP::StopEventHandlers() {
+  // as parameter to `SendJSON`.
+  protocol::Message M;
+  llvm::json::Path::Root root;
+  if (!protocol::fromJSON(json, M, root)) {
```
Now that I see how this function is used, wouldn't it be better for `protocol::fromJSON` to return an `std::optional<protocol::Message>` instead of returning a bool and taking the message by reference?

JDevlieghere commented on lldb/tools/lldb-dap/DAP.cpp:
```diff
@@ -219,65 +220,27 @@ void DAP::StopEventHandlers() {
-         << "Content-Length: " << length << "\r\n\r\n";
+  // FIXME: Instead of parsing the output message from JSON, pass the `Message`
+  // as parameter to `SendJSON`.
+  protocol::Message M;
```
This patch is using a mix of LLVM and LLDB naming conventions. Let's be consistent and use lower_camel_case.

JDevlieghere commented on lldb/tools/lldb-dap/Transport.cpp:
```diff
@@ -0,0 +1,146 @@
+    return std::move(Err);
+  if (rawJSON->length() != length)
+    return createStringError(
+        "malformed request, expected %ld bytes, got %ld bytes", length,
```
We have a few places in lldb which concatenate errors with `:` so I'm trying to be more consistent in using coons to separate error messages. You used a colon on line 120 so let's do the same here. 

```suggestion
        "malformed request: expected %ld bytes and got %ld bytes", length,
```

labath replied:
Also, `%ld` is not a correct way to print a variable of type `size_t`. It's `%zd`, but I think it'd be even better to avoid this question and format these strings with `llvm::formatv`, even though the result is a bit longer.

JDevlieghere commented on lldb/tools/lldb-dap/Transport.cpp:
```diff
@@ -0,0 +1,146 @@
+
+  auto JSON = json::parse(*rawJSON);
+  if (auto Err = JSON.takeError()) {
+    return createStringError("malformed JSON %s\n%s", rawJSON->c_str(),
+                             llvm::toString(std::move(Err)).c_str());
```
It's probably worth adding a colon here too:

```suggestion
    return createStringError("malformed JSON: %s\n%s", rawJSON->c_str(),
                             llvm::toString(std::move(Err)).c_str());
```

Which makes me wonder: is it worth including the raw JSON in the error? How about returning just the error and logging the raw JSON?

JDevlieghere reviewed (commented):
I left some small comments but overall I really like this approach.

ashgti committed:
Adjusting `Transport::Read` to return an `optional<string>` and adding logging directly to the `Transport` class.

labath commented on lldb/tools/lldb-dap/DAP.cpp:
```diff
@@ -838,19 +775,16 @@ llvm::Error DAP::Loop() {
-      return llvm::createStringError(llvm::inconvertibleErrorCode(),
-                                     "failed to send packet");
+    auto next = transport.Read(log);
+    if (auto Err = next.takeError()) {
+      // On EOF, simply break out of the loop.
+      std::error_code ec = llvm::errorToErrorCode(std::move(Err));
+      if (ec == Transport::kEOF)
+        break;
+      return llvm::errorCodeToError(ec);
     }
```
Then this would be something like:
```suggestion
    if (auto Err = next.takeError()) {
      // On EOF, simply break out of the loop.
      return handleErrors(std::move(Err), [](const EOFError &) {});
    }
```

ashgti replied:
I tweaked the `Transport` class to log any errors it encounters and return `std::nullopt` whenever there is a problem, including EOF.

labath commented on lldb/tools/lldb-dap/Transport.cpp:
```diff
@@ -0,0 +1,146 @@
+
+  // If we got back zero then we have reached EOF.
+  if (length == 0)
+    return createStringError(Transport::kEOF, "end-of-file");
```
```suggestion
    return makeError<EOFError>();
```

ashgti replied:
I tweaked the `Transport` class to log any errors it encounters and return `std::nullopt` whenever there is a problem, including EOF.

labath commented on lldb/tools/lldb-dap/Transport.cpp:
```diff
@@ -0,0 +1,146 @@
+using namespace lldb_dap;
+using namespace lldb_dap::protocol;
+
+static Expected<std::string> ReadFull(IOObjectSP &descriptor, size_t length) {
```
A (non-const) reference to a shared pointer usually not what you want. Since this function doesn't really need the shared-pointerness. I'd go with a reference to the underlying object instead.
```suggestion
static Expected<std::string> ReadFull(IOObject &descriptor, size_t length) {
```

ashgti replied:
Tweaked to use a pointer to the `IOObject *` (which is abstract so I can't use a reference to the base class).

labath replied:
That probably fine, but I am pretty sure that being abstract doesn't prevent you from [forming a reference](https://godbolt.org/z/Ejcxve5qq) to a class. You must have tried to copy the object somewhere. The main advantage of a reference is that it guarantees that the pointer is not null.

labath commented on lldb/tools/lldb-dap/Transport.cpp:
```diff
@@ -0,0 +1,146 @@
+  if (auto Err = ReadExpected(m_input, "Content-Length: "))
+    return std::move(Err);
+
+  auto rawLength = ReadUntil(m_input, "\r\n\r\n");
```
Is there any risk in this appearing in the middle of a single block returned by the read call (e.g. if the transport layer merges the `\n` terminating one request with the subsequent request)?

ashgti replied:
Shouldn't break the output, even though it will produce something that looks like a message in the middle of a message.

We're not searching for the `Content-Length: ` header, we expect it to be those exact bytes, per the spec.  The messages should always be formatted like:

```
Content-Length: (\d+)\r\n\r\n(.{\1})
```

Or more generally:

```
<header><length><separator><body>

  <header> = "Content-Length: "
  <length> = \d+
  <separator> = "\r\n\r\n"
  <body> = length bytes
```

Otherwise the message is malformed.

Once we have a message header, if the `Content-Length: ` string appears in the body of the message, that should be okay. 
e.g.

```
int main() {
  printf("Content-Length: 51\r\n\r\n{\"command\":\"disconnect\",\"seq\": 1,\"type\": \"request\"}");
  return 0;
}
```

Works fine with VSCode and doesn't have any problems with lldb-dap.

labath replied:
Okay, I see what you're doing now. `ReadUntil` is reading the string one character at a time, which means it can never read past the \r\n terminator. Not particularly efficient, but I suppose it will do given that this just needs to read a couple of bytes.

ashgti commented on lldb/tools/lldb-dap/Transport.cpp:
```diff
@@ -0,0 +1,146 @@
+namespace lldb_dap {
+
+const std::error_code Transport::kEOF =
+    std::error_code(0x1001, std::generic_category());
```
I ended up refactoring this to use an `std::optional<protocol::Message>` instead of a `llvm::Expected<protocol::Message>`.

I don't think returning the error was actually helpful and the error handling was getting complicated due to treating EOF as an error.

Instead, I tweaked the `Transport` class to log any errors it encounters and return `std::nullopt` whenever there is a problem, including EOF.

labath replied:
Hmm... I can't say I like the repetitiveness of of the log statements in the Read function, nor the fact that `Read` and `Write` methods handle errors differently (one logs while the other returns them). What exactly was making that complicated?

Another option would be to return an `Expected<optional<Message>>` with a `nullopt` meaning "EOF". It's a bit of a mouthful, but we do have APIs like that, and it doesn't look like this function will be used from that many places.

ashgti committed:
Adjusting the reader helpers to take `lldb_private::IOObject *` instead of the shared_ptr.

ashgti committed:
Aplying clang-format.

labath commented on lldb/tools/lldb-dap/DAP.cpp:
```diff
@@ -838,19 +770,12 @@ llvm::Error DAP::Loop() {
-
-    if (status == lldb_dap::PacketStatus::EndOfFile) {
+    std::optional<protocol::Message> next = transport.Read();
+    if (!next) {
       break;
     }
```
```suggestion
    if (!next)
      break;
```

ashgti replied:
Done.

ashgti commented on lldb/tools/lldb-dap/DAPLog.h:
```diff
@@ -0,0 +1,58 @@
+}
+} // namespace lldb_dap
+
+#endif
```
Fixed in the baseline #130653

ashgti committed:
Applying reviewer feedback.

ashgti commented on lldb/tools/lldb-dap/lldb-dap.cpp:
```diff
@@ -315,23 +316,19 @@ serveConnection(const Socket::SocketProtocol &protocol, const std::string &name,
-                    pre_init_commands);
+      llvm::set_thread_name(client_name + ".runloop");
+      Transport transport{client_name, log, io, io};
+      DAP dap = DAP(program_path, log, default_repl_mode, pre_init_commands,
```
Updated to c++11 basic style for consistency

ashgti commented on lldb/tools/lldb-dap/Transport.cpp:
````diff
@@ -0,0 +1,144 @@
+/// Content-Length: (?<length>\d+)\r\n\r\n(?<content>.{\k<length>})
+/// ```
+static const StringLiteral kHeaderContentLength = "Content-Length: ";
+static const StringLiteral kHeaderSeparator = "\r\n\r\n";
````
Done.

ashgti committed:
Adjusting `Transport.Read` to return an `llvm::Expected<std::optional<protocol::Message>>` to try to refine the caller.

An error reading or validating a message should now result in an `llvm::Error`.

EOF is now only acceptable at the start of a message, otherwise we return an error indiciating we encountered a partial message.

labath commented on lldb/tools/lldb-dap/DAP.cpp:
```diff
@@ -63,12 +64,12 @@ const char DEV_NULL[] = "/dev/null";
+         llvm::StringRef client_name, Transport &transport)
+    : debug_adapter_path(path), log(log), client_name(client_name),
+      transport(transport), broadcaster("lldb-dap"), exception_breakpoints(),
       pre_init_commands(std::move(pre_init_commands)),
```
This `std::move` us a noop now that the argument is `const&`. if you want to keep it that way, then delete the move, though I actually think the original code was better as it lets the caller decide whether it needs its own copy of the argument or not (where as this guarantees there will be at least one copy operation)

ashgti replied:
Reverted `pre_init_commands` changes

labath reviewed (approved):
LGTM

JDevlieghere commented on lldb/tools/lldb-dap/DAP.cpp:
```diff
@@ -221,52 +222,22 @@ void DAP::StopEventHandlers() {
+                  client_name);
+    return;
+  }
+  auto err = transport.Write(message);
+  if (err) {
+    DAP_LOG_ERROR(log, std::move(err), "({1}) write failed: {0}", client_name);
+  }
```
Nit: the type of `err` is not obvious from the RHS so I don't think `auto` is appropriate. Also no braces for single line `if`s:
```suggestion
  if (llvm::Error err = transport.Write(message)) 
    DAP_LOG_ERROR(log, std::move(err), "({1}) write failed: {0}", client_name);
  ```

JDevlieghere commented on lldb/tools/lldb-dap/DAP.h:
```diff
@@ -210,12 +209,33 @@ struct DAP {
+  /// \param[in] pre_init_commands
+  ///     LLDB commands to execute as soon as the debugger instance is allocaed.
+  /// \param[in] client_name
+  ///     Debug session client name, for example 'stdin/stdout' or 'client_1'.
```
What's the reason for moving the `path` and `client_name` apart?

JDevlieghere replied:
If the `client_name` is stored in the `Transport` class, do we need to duplicate it in the `DAP` object? Does it (not) make sense to get it from the transport instance?

ashgti replied:
> What's the reason for moving the `path` and `client_name` apart?

I tried to reorganize the parameters into more constant between DAP sessions to more unique per DAP session. i.e. 

* `path` is basically a constant (its the path to `lldb-dap` itself)
* `log` is also either setup or not at launch
* `default_repl_mode` and `pre_init_commands` are CLI flags that affect the starting conditions of the DAP session
* `client_name` and `transport` are the most unique values for the DAP session.

> If the `client_name` is stored in the `Transport` class, do we need to duplicate it in the `DAP` object? Does it (not) make sense to get it from the transport instance?

I'll update this to access the client name from the transport.

JDevlieghere commented on lldb/tools/lldb-dap/Transport.cpp:
```diff
@@ -0,0 +1,126 @@
+using namespace lldb_dap::protocol;
+
+/// ReadFull attempts to read the specified number of bytes. If EOF is
+/// encountered, '' is returned.
```
```suggestion
/// encountered, an empty string is returned.
```

JDevlieghere commented on lldb/tools/lldb-dap/lldb-dap.cpp:
```diff
@@ -565,8 +568,10 @@ int main(int argc, char *argv[]) {
 
-  DAP dap = DAP("stdin/stdout", program_path, log.get(), std::move(input),
-                std::move(output), default_repl_mode, pre_init_commands);
+  std::string client_name = "stdin/stdout";
```
```suggestion
  llvm::StringLiteral client_name = "stdin/stdout";
```

JDevlieghere reviewed (approved):
LGTM modulo nits.

ashgti committed:
Apply suggestions from code review

Co-authored-by: Jonas Devlieghere <jonas@devlieghere.com>

ashgti committed:
Removing DAP::client_name and instead only storing the name in Transport

ashgti committed:
Reverting changes to `pre_init_commands` type in DAP::DAP.

ashgti committed:
Applying clang-format.
